### PR TITLE
Add test for cross-contract reentrancy

### PR DESCRIPTION
**Pull request by nfurfaro**
## Type of change

- Improvement (refactoring, restructuring repository, cleaning tech debt, ...)

## Changes

The following changes have been made:

- added a new test case
- added supporting new contract & abi
- added new functions on existing contracts & abis

## Notes

The test case added performs the following steps:

1. The `Attack` contract's `launch_thwarted_attack_3()` function is triggered by a contract call from the SDK.
2. The `Attack` contract calls into the `Target` contract's `cross_contract_reentrancy_denied()` function.
3. The `Target` contract calls back to the `Attack` contract's `evil_callback_4()` function.
4. The `Attack` contract calls into the `Attack Helper` contract's `attempt_cross_contract_reentrancy()` function.
5. (attack vector) The `Attack Helper` contract attempts to re-enter by calling into the `Target` contract's `cross_contract_reentrancy_denied()` function(which hasn't finished execution of the initial call from the `Attack` contract).
6.  The attack is thwarted by the `reentrancy_guard()` function in place.

### CODE REVIEW

File: tests/src/reentrancy/reentrancy_attack_helper_contract/Forc.toml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+[project]
+authors = ["   Nicholas Furfaro"]
```
Comment (Braqzen) [inline tests/src/reentrancy/reentrancy_attack_helper_contract/Forc.toml:2]:
```suggestion
authors = ["Nicholas Furfaro"]
```

**Review by Braqzen — APPROVED** [review]
Seems fine but is there something preventing a 2 contract interaction over 3 contracts?
Contract A calls into Contract B (non-reentrant func)
Contract B calls into Contract A
Contract A calls back into Contract B (non-reentrant func)

**Review by pixelcircuits — APPROVED** [review]

File: tests/src/reentrancy/reentrancy_attack_helper_contract/Forc.toml (hunk 1 above)
Comment (nfurfaro) [reply tests/src/reentrancy/reentrancy_attack_helper_contract/Forc.toml]: fixed.

**Comment by nfurfaro** [comment]
>Seems fine but is there something preventing a 2 contract interaction over 3 contracts?

@Braqzen I'm not quite sure what you're getting at here.
If the `reentrancy_guard()` function is _not_ used, given 2 Contracts `A` & `B`:
- `A` can call into `B` can call into `A` can call into `B` can call into `A` can call into `B` ... until the gas runs out.
- The compiler just warns us if the CEI (Checks, Effects, Interactions) pattern is violated. It doesn't strictly prevent reentrant function calls.

**Comment by Braqzen** [comment]
> > Seems fine but is there something preventing a 2 contract interaction over 3 contracts?
> 
> @Braqzen I'm not quite sure what you're getting at here. If the `reentrancy_guard()` function is _not_ used, given 2 Contracts `A` & `B`:
> 
> * `A` can call into `B` can call into `A` can call into `B` can call into `A` can call into `B` ... until the gas runs out.
> * The compiler just warns us if the CEI (Checks, Effects, Interactions) pattern is violated. It doesn't strictly prevent reentrant function calls.

By "non-reentrant func" I meant to use the guard in that function, I should have been clearer, my mistake.

**Review by K1-R1 — APPROVED** [review]

**Pull request merged**